### PR TITLE
Prefer optional chain

### PR DESCRIPTION
**Pull request by marek-saji**
# Why? <a id=user-content-why href=#user-content-why>🔗</a>

I just wanted to add a plugin with a rule for preferring optional chaining (`foo?.bar?.zar` over `foo && foo.bar && foo.bar.zar`).

# What? <a id=user-content-what href=#user-content-what>🔗</a>

Ended up doing few things more:

1. README erroneously said that it’s enough to install this package and you are ready to go. But — similar to AirBnB config, which we base on — you need all the plugins that we are using.
   I copied AirBnB’s approach and duplicated those dependencies as `peerDependencies` and added info on how to install them in the README.

2. Since we are starting to use TypeScript I’ve added `.ts` and `.tsx` extensions.

3. Updated order of things in classes to place static properties before methods.

4. Removed some of the rules that are conflicting with prettier.

5. Speaking of prettier, I included prettier configuration in this package with instructions on how to use it.

6. Package now uses it’s own configuration to lint itself.

And yeah, I’ve added `@typescript-eslint/eslint-plugin`, which gives us the `prefer-optional-chain` rule, but that’s not enabled by default, since project using this config may not have it set up.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -24,18 +23,31 @@ module.exports = {
...
     'react/jsx-wrap-multilines': 'off',
     'react/prefer-stateless-function': 'off',
-    'react/sort-comp': ['error', { order: ['static-methods', 'instance-variables', 'lifecycle', 'everything-else', 'render'] }],
+    'react/sort-comp': [
+      'error',
+      {
+        order: [
+          'static-variables',
+          'instance-variables',
+          'static-methods',
```
Comment (josipovich) [inline index.js:44]: maybe to hoist statics first?

**Comment by josipovich** [comment]
> I’ve added @typescript-eslint/eslint-plugin, which gives us the prefer-optional-chain rule, but that’s not enabled by default, since project using this config may not have it set up.

 ~~I guess this can be done in fock's eslint config then. Also might be worth mentioning that depending on the errors count we could make it warning for first time so we don't need to refactor.~~

Edit: wrote this before I found out about fock part of the PR

File: index.js (hunk 1 above)
Comment (marek-saji) [reply index.js:44]: Do you want static methods before static variables? I’d prefer having variables first as they are usually shorter.

Comment (josipovich) [reply index.js:44]: Hmm, I think you make a good point! All good, ignore me :D

**Comment by marek-saji** [comment]
> I guess this can be done in fock's eslint config then. Also might be worth mentioning that depending on the errors count we could make it warning for first time so we don't need to refactor.

Doesn’t make sense to keep it as a warning for projects that don’t plan on supporting that operator (e.g. don’t do any transpiling). … But if most of our projects do plan to support chaining, we can enable it here and let package maintainers disable it where it’s not needed.

File: prettier-config.json [hunk 2]
```diff
@@ -0,0 +1,10 @@
+{
+  "endOfLine": "lf",
+  "printWidth": 80,
+  "semi": false,
+  "singleQuote": true,
+  "quoteProps": "preserve",
+  "tabWidth": 2,
+  "TRAIlingComma": "none",
```
Comment (nuc) [inline prettier-config.json]: Uppercasing intensifies 😄

**Review by nuc — APPROVED** [review]

File: prettier-config.json (hunk 2 above)
Comment (marek-saji) [reply prettier-config.json]: Ugh. Fixed.

**Review by josipovich — APPROVED** [review]

**Pull request merged**